be/src/exec/vectorized/chunk_sorter_heapsorter.cpp

```diff
@@ -0,0 +1,239 @@
+            for (; i < direct_push; ++i) {
+                detail::ChunkRowCursor cursor(i, chunk_holder);
+                _sort_heap->push(std::move(cursor));
+            }
+
+            // compare to heap top and replace top
+            for (; i < row_sz; ++i) {
+                detail::ChunkRowCursor cursor(i, chunk_holder);
+                _sort_heap->replace_top_if_less(std::move(cursor));
+            }
+        }
+    }
+    // TODO: merge chunk if necessary
+    return Status::OK();
+}
+
+Status HeapChunkSorter::done(RuntimeState* state) {
+    ScopedTimer<MonotonicStopWatch> timer(_build_timer);
+    if (_sort_heap) {
+        _sorted_values = _sort_heap->sorted_seq();
+    }
+    _sort_heap.reset();
+    return Status::OK();
+}
+
+void HeapChunkSorter::get_next(ChunkPtr* chunk, bool* eos) {
+    ScopedTimer<MonotonicStopWatch> timer(_output_timer);
+
+    if (_next_output_row == _sorted_values.size()) {
+        *eos = true;
+        *chunk = nullptr;
+        return;
+    }
+
+    const size_t count = std::min(size_t(config::vector_chunk_size), _sorted_values.size() - _next_output_row);
+    chunk->reset(_sorted_values[0].data_segment()->chunk->clone_empty(count).release());
+
+    for (int i = _next_output_row; i < count + _next_output_row; ++i) {
+        auto rid = _sorted_values[i].row_id();
+        const auto& ref_chunk = _sorted_values[i].data_segment()->chunk;
+        (*chunk)->append_safe(*ref_chunk, rid, 1);
+    }
+
+    _next_output_row += count;
+}
+
+template <PrimitiveType TYPE>
+void HeapChunkSorter::_do_filter_data_for_type(detail::ChunkHolder* chunk_holder, Column::Filter* filter, int row_sz) {
+    const auto& top_cursor = _sort_heap->top();
+    const int cursor_rid = top_cursor.row_id();
+
+    const auto& top_cursor_column = top_cursor.data_segment()->order_by_columns[0];
+    const auto& input_column = chunk_holder->value()->order_by_columns[0];
+
+    DCHECK_EQ(top_cursor_column->is_nullable(), input_column->is_nullable());
+
+    if (top_cursor_column->is_nullable()) {
+        bool top_is_null = top_cursor_column->is_null(cursor_rid);
+        const auto& need_filter_data =
+                ColumnHelper::cast_to_raw<TYPE>(down_cast<NullableColumn*>(top_cursor_column.get())->data_column())
+                        ->get_data()[cursor_rid];
+
+        const auto& order_by_null_column = down_cast<NullableColumn*>(input_column.get())->null_column();
+        const auto& order_by_data_column = down_cast<NullableColumn*>(input_column.get())->data_column();
+
+        const auto* null_data = order_by_null_column->get_data().data();
+        const auto* order_by_data = ColumnHelper::cast_to_raw<TYPE>(order_by_data_column)->get_data().data();
+        auto* __restrict__ filter_data = filter->data();
+
+        for (int i = 0; i < row_sz; ++i) {
+            // data is null
+            if (null_data[i] && !top_is_null) {
+                filter_data[i] = _null_first_flag[0] * _sort_order_flag[0] < 0;
```

Comment:
Could use &, not * ?

Reply:
we could't use &, but we could recode _null_first_flag and _sort_order_flag to [1, 0], then we could use xor instead of `*`, but it's too hard to understand.